server/src/main/java/io/druid/segment/realtime/RealtimeManager.java

```diff
@@ -238,67 +261,18 @@ public FireDepartmentMetrics getMetrics()
     public void run()
     {
       plumber = initPlumber();
-      final Period intermediatePersistPeriod = config.getIntermediatePersistPeriod();
 
       try {
-        plumber.startJob();
+        Object metadata = plumber.startJob();
 
-        // Delay firehose connection to avoid claiming input resources while the plumber is starting up.
-        firehose = initFirehose();
-
-        long nextFlush = new DateTime().plus(intermediatePersistPeriod).getMillis();
-        while (firehose.hasMore()) {
-          InputRow inputRow = null;
-          try {
-            try {
-              inputRow = firehose.nextRow();
-
-              if (inputRow == null) {
-                log.debug("thrown away null input row, considering unparseable");
-                metrics.incrementUnparseable();
-                continue;
-              }
-            }
-            catch (ParseException e) {
-              log.debug(e, "thrown away line due to exception, considering unparseable");
-              metrics.incrementUnparseable();
-              continue;
-            }
-
-            boolean lateEvent = false;
-            boolean indexLimitExceeded = false;
-            try {
-              lateEvent = plumber.add(inputRow) == -1;
-            }
-            catch (IndexSizeExceededException e) {
-              log.info("Index limit exceeded: %s", e.getMessage());
-              indexLimitExceeded = true;
-            }
-            if (indexLimitExceeded || lateEvent) {
-              metrics.incrementThrownAway();
-              log.debug("Throwing away event[%s]", inputRow);
-
-              if (indexLimitExceeded || System.currentTimeMillis() > nextFlush) {
-                plumber.persist(firehose.commit());
-                nextFlush = new DateTime().plus(intermediatePersistPeriod).getMillis();
-              }
-
-              continue;
-            }
-            final Sink sink = plumber.getSink(inputRow.getTimestampFromEpoch());
-            if ((sink != null && !sink.canAppendRow()) || System.currentTimeMillis() > nextFlush) {
-              plumber.persist(firehose.commit());
-              nextFlush = new DateTime().plus(intermediatePersistPeriod).getMillis();
-            }
-            metrics.incrementProcessed();
-          }
-          catch (ParseException e) {
-            if (inputRow != null) {
-              log.error(e, "unparseable line: %s", inputRow);
-            }
-            metrics.incrementUnparseable();
-          }
+        if (fireDepartment.checkFirehoseV2()) {
+          firehoseV2 = initFirehose(metadata);
+          runFirehoseV2(firehoseV2);
+        } else {
+          firehose = initFirehose();
+          runFirehose(firehose);
```

Comment:
Is it possible to wrap a Firehose and present it as a FirehoseV2? I thought that was one of the goals of FirehoseV2. That would make this code a lot cleaner, since it would only be dealing with FirehoseV2s.